### PR TITLE
Implement the Astro.request RFC

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Implements the `Astro.request` RFC changes: https://github.com/withastro/rfcs/blob/main/proposals/0018-astro-request.md
- Warns on usage of `Astro.request.params` and `Astro.request.canonicalURL`.

## Testing

Tests updated

## Docs

- https://github.com/withastro/docs/pull/270

### CODE REVIEW

File: packages/astro/src/@types/astro.ts [hunk 1]
```diff
@@ -47,10 +46,14 @@ export interface BuildConfig {
...
 export interface AstroGlobal extends AstroGlobalPartial {
+	/** get the current canonical URL */
+	canonicalURL: URL;
+	/** get page params (dynamic pages only) */
+	params: Params;
 	/** set props for this astro component (along with default values) */
 	props: Record<string, number | string | any>;
 	/** get information about this page */
-	request: AstroRequest;
+	request: Request;
```
Comment (bholmesdev) [inline packages/astro/src/@types/astro.ts:56]: _love_ this change 👏

File: packages/astro/src/core/request.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	
+	const request = new Request(url.toString(), {
+		method: method,
+		headers: headersObj
+	});
+
+	Object.defineProperties(request, {
+		canonicalURL: {
+			get() {
+				console.warn(`Astro.request.canonicalURL has been moved to Astro.canonicalURL`);
```
Comment (bholmesdev) [inline packages/astro/src/core/request.ts:17]: Nit: should we be using `logger` instead here? Not certain

**Review by bholmesdev — APPROVED** [review]
Left one `request`ion about logging, but Let's GET this money 💰

File: packages/astro/src/core/request.ts (hunk 2 above)
Comment (matthewp) [reply packages/astro/src/core/request.ts:17]: Yeah I guess we should, I'm not sure why I did it this way. Can change.

File: packages/astro/src/@types/astro.ts (hunk 1 above)
Comment (FredKSchott) [reply packages/astro/src/@types/astro.ts:59]: seriously, it's the little things :)

File: packages/astro/src/vite-plugin-astro-server/index.ts [hunk 3]
```diff
@@ -117,9 +117,24 @@ async function handleRequest(
...
 	const rootRelativeUrl = pathname.substring(devRoot.length - 1);
+	if(!buildingToSSR) {
+		// Prevent user from depending on search params when not doing SSR.
+		for(const [key] of url.searchParams) {
+			url.searchParams.delete(key);
+		}
+	}
+
+	// Headers are only available when using SSR.
+	const request = createRequest({
```
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-astro-server/index.ts:132]: TIL that Node.js `http.IncomingMessage` wasn't a `Request` object? For some reason I thought that that had always been the case.

**Review by FredKSchott — APPROVED** [review]

**Comment by FredKSchott** [comment]
Did my best to fix merge conflicts for you, but feel free to revert [c6b9c69](https://github.com/withastro/astro/pull/2913/commits/c6b9c691c39ba3d3f1f067d18368fddd7b7ef8a7) and do your own rebase if it causes trouble for you

**Comment by FredKSchott** [comment]
Did my best to fix merge conflicts for you. If you need to, you can force-push your local branch back up github to revert my fixes.

**Pull request merged**